HIVE-25667: Unify code managing JDBC databases in tests

mbathori-cloudera opened the pull request:
### What changes were proposed in this pull request?
Currently there are two class hierarchies managing JDBC databases in tests, DatabaseRule and AbstractExternalDB. These changes meant to unify these database related functionalities and get rid of the duplicated codes.


### Why are the changes needed?
The current solution is redundant, contains a lot of code duplications. The unified database handling gives easier extensibility and modification.


### Does this PR introduce _any_ user-facing change?
No


### How was this patch tested?
All the previous tests that were covering the affected database functionality are passing with the same expected results.

zabetak commented on standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/dbinstall/AbstractDatabase.java:
```diff
@@ -0,0 +1,302 @@
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+public abstract class AbstractDatabase extends ExternalResource {
```
Do we want to keep this dependent to `ExternalResource` and junit APis in general?

zabetak commented on standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/dbinstall/AbstractDatabase.java:
```diff
@@ -0,0 +1,302 @@
+package org.apache.hadoop.hive.metastore.dbinstall;
```
Isn't `hive-testutils` module a better place for keeping these classes since they are not only used in the `metastore-server`?

mbathori-cloudera replied:
Yes I felt that the metastore isn’t the right place for this class but didn't know where should it be, I’ll move it into the hive-testutils module.

zabetak replied:
The `hive-testutils` module seems to be the right place for this classes we should be careful though not to create circular dependencies between hive-testutils and metastore modules.

zabetak commented on standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/dbinstall/AbstractDatabase.java:
```diff
@@ -0,0 +1,302 @@
+        "-userName", getHiveUser(),
+        "-passWord", getHivePassword(),
+        "-url", getJdbcUrl(),
+        "-driver", getJdbcDriver(),
+        "-verbose"
+    });
+  }
+
+  public int installAVersion(String version) {
+    return new MetastoreSchemaTool().setVerbose(verbose).run(new String[] {
+        "-initSchemaTo", version,
+        "-dbType", getDbType(),
+        "-userName", getHiveUser(),
+        "-passWord", getHivePassword(),
+        "-url", getJdbcUrl(),
+        "-driver", getJdbcDriver()
+    });
+  }
+
+  public int upgradeToLatest() {
+    return new MetastoreSchemaTool().setVerbose(verbose).run(new String[] {
+        "-upgradeSchema",
+        "-dbType", getDbType(),
+        "-userName", getHiveUser(),
+        "-passWord", getHivePassword(),
+        "-url", getJdbcUrl(),
+        "-driver", getJdbcDriver()
+    });
+  }
+
+  public int validateSchema() {
+    return new MetastoreSchemaTool().setVerbose(verbose).run(new String[] {
+        "-validate",
+        "-dbType", getDbType(),
+        "-userName", getHiveUser(),
+        "-passWord", getHivePassword(),
+        "-url", getJdbcUrl(),
+        "-driver", getJdbcDriver()
+    });
+  }
```
Possibly we could do this in separate refactoring/jira but these methods are not necessary for all consumers of this API so we could put them in a separate class/interface.

zabetak commented on standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/dbinstall/rules/Mssql.java:
```diff
@@ -63,7 +64,8 @@ public String getJdbcDriver() {
   @Override
   public String getJdbcUrl(String hostAddress) {
-    return "jdbc:sqlserver://" + hostAddress + ":1433;DatabaseName=" + HIVE_DB + ";";
+    //Mssql doesn't support database parameter in docker
```
I don't understand what the comment means. Can you elaborate more?

mbathori-cloudera replied:
The Mssql docker init doesn’t have database parameter for initialisation like postgres/mysql etc. so the initial jdbc connection should go to default url if the database was created with docker database argument.(https://github.com/microsoft/mssql-docker/issues/2)

zabetak commented on standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/dbinstall/AbstractDatabase.java:
```diff
@@ -0,0 +1,302 @@
+    }
+  }
+
+  @Override
+  public void before() throws Exception {
+    launchDockerContainer();
+    MetastoreSchemaTool.setHomeDirForTesting();
+  }
+
+  @Override
+  public void after() {
+    if ("true".equalsIgnoreCase(System.getProperty("metastore.itest.no.stop.container"))) {
+      LOG.warn("Not stopping container " + getDockerContainerName() + " at user request, please "
+          + "be sure to shut it down before rerunning the test.");
+      return;
+    }
+    try {
+      cleanupDockerContainer();
+    } catch (InterruptedException | IOException e) {
+      e.printStackTrace();
+    }
+  }
```
Possibly these methods can go to those classes using this class as a rule. Another alternative would be to create a rule class which wraps an instance of `AbstractDatabase` and provides these methods.

zabetak commented on standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/dbinstall/rules/Mariadb.java:
```diff
@@ -18,19 +18,31 @@
+
+  @Override
+  public List<String> getDockerBaseArgs() {
+    return new ArrayList(Arrays.asList("-p", "3309:3306",
```
Why are we wrapping into ArrayList? Are we going to modify the list? Keeping things immutable makes the code easier to understand and use.

mbathori-cloudera replied:
I was wrapping this because further database argumentum can be added in the AbstractDatabase class, so the list can’t be immutable. If we are getting rid of the docker database argument this will be removed.

zabetak replied:
I think that we could create a new list inside `AbstractDatabse.getDockerAdditionalArgs()` if necessary instead of modifying those that we obtain from the getters.

zabetak commented on standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/dbinstall/rules/Derby.java:
```diff
@@ -42,7 +46,7 @@ public String getDockerImageName() {
   @Override
-  public String[] getDockerAdditionalArgs() {
+  public List<String> getDockerBaseArgs() {
     return null;
```
Effective Java Item 43: Return empty arrays or collections, not nulls

zabetak commented on standalone-metastore/metastore-server/src/test/java/org/apache/hadoop/hive/metastore/dbinstall/AbstractDatabase.java:
```diff
@@ -0,0 +1,302 @@
+    return String.format("testDb-%s", getClass().getSimpleName());
+  }
+
+  private List<String> getDockerAdditionalArgs() {
+    List<String> dockerArgs = getDockerBaseArgs();
+    if (useDockerDatabaseArg && StringUtils.isNotEmpty(getDockerDatabaseArg())) {
+      dockerArgs.addAll(Arrays.asList("-e", getDockerDatabaseArg()));
+    }
+    return dockerArgs;
+  }
```
Having many `getDocker*Args` method makes the code harder to follow. Does it bring a big benefit having  `getDockerBaseArgs and `getDockerDatabaseArg` separate?

mbathori-cloudera replied:
Currently every database implementation has a base argument list for docker initialisation, but not all of them has database arguments (eg. Mysql,Oracle). I wanted to make it configurable so we can initialise the database with docker like it was in` AbstractExternalDB.java` or we can create the database in the init/test script like in `DatabaseRule.java`. I wasn’t really sure if I should use one of these approach for both so I kept their original functionality.
What do you think about this?

zabetak replied:
OK makes sense. Thanks for the clarification.

zabetak reviewed (commented):
Many thanks for picking this up @mbathori-cloudera ! I went quickly over the PR and left a few comments.

mbathori-cloudera commented:
@zabetak Thanks for reviewing my changes. I made a new commit where I tried to address all of your comments.
- I moved the common database classes to the `hive-testutils` module.
- Moved metastore related parts from common database classes to a wrapper class in the metastore.
- Removed the `ExternalResource` and refactored depending functions.


**_Some note for the hive-testutil pom changes:_**
I had a strange issue while adding `hive-testutils` to the `itests/util/pom.xml`. The qtests were not executed and the build automatically skipped all of them without any error message. It turned out that a dependency called `junit-jupiter-api` caused this issue. If I set this dependency to test scope the problem would be solved but I can’t do it because classes that are using it are placed in the src package. The easiest solution was to set the dependency to _optional_ so it will be excluded.

mbathori-cloudera commented:
It seems like the metastore verify stage is failing in the pipeline because I refactored the DbInstall tests and it still tries to run the ITestDerby,ITestMysql  and ITestPostgres tests but they don’t exist anymore. All of these test are now running in ITestDbInstall. I’m not sure if the pipeline needs to be adjusted to the changes or should I restore the original form of these tests.

zabetak commented:
Hey bot please keep this PR open!
